### PR TITLE
Backwards compatibility

### PR DESCRIPTION
**Pull request by vjosadika**
JIRA ticket: https://issues.cask.co/browse/PLUGIN-180

### CODE REVIEW

File: database-plugins/src/main/java/io/cdap/plugin/DBRecord.java [hunk 1]
```diff
@@ -99,7 +101,18 @@ public StructuredRecord getRecord() {
...
     ResultSetMetaData metadata = resultSet.getMetaData();
-    List<Schema.Field> originalSchema = DBUtils.getOriginalSchema(resultSet);
+    String outputSchemaString = conf.get(DBUtils.OVERRIDE_SCHEMA, null);
+    Schema outputSchema = null;
+
+    if (!Strings.isNullOrEmpty(outputSchemaString)) {
+      try {
+        outputSchema = Schema.parseJson(outputSchemaString);
+      } catch (IOException e) {
+        e.printStackTrace();
```
Comment (CuriousVini) [inline database-plugins/src/main/java/io/cdap/plugin/DBRecord.java]: we shouldnt do it. the exception should be logged. Please take a look at other places where we handle this case in the codebase

File: database-plugins/src/main/java/io/cdap/plugin/db/batch/source/DBSource.java [hunk 2]
```diff
@@ -318,20 +319,20 @@ private Connection getConnection() throws SQLException {
     @Nullable
     @Name(PATTERN_TO_REPLACE)
     @Description("The pattern to replace in the field name in the table, it is typically used with the " +
-                   "Replace With config. If Replace With is not set, the pattern will be removed in the field name.")
+      "Replace With config. If Replace With is not set, the pattern will be removed in the field name.")
```
Comment (CuriousVini) [inline database-plugins/src/main/java/io/cdap/plugin/db/batch/source/DBSource.java]: these changes are not needed. please revert them

File: database-plugins/src/main/java/io/cdap/plugin/DBRecord.java [hunk 3]
```diff
@@ -283,7 +312,7 @@ private void writeToDB(PreparedStatement stmt, Schema.Field field, int fieldInde
         break;
       default:
         throw new SQLException(String.format("Column %s with value %s has an unsupported datatype %s",
-          field.getName(), fieldValue, fieldType));
+                                             field.getName(), fieldValue, fieldType));
```
Comment (CuriousVini) [inline database-plugins/src/main/java/io/cdap/plugin/DBRecord.java]: these changes are not needed . Please revert them

File: database-plugins/src/main/java/io/cdap/plugin/DBUtils.java [hunk 4]
```diff
@@ -232,13 +246,23 @@ private static Schema getSchema(String typeName, int sqlType, int precision, int
         break;
 
       case Types.NUMERIC:
-      case Types.DECIMAL:
+        //variation
```
Comment (CuriousVini) [inline database-plugins/src/main/java/io/cdap/plugin/DBUtils.java]: not needed. Please remove.

File: database-plugins/src/main/java/io/cdap/plugin/DBRecord.java [hunk 5]
```diff
@@ -124,23 +137,39 @@ public void readFields(ResultSet resultSet) throws SQLException {
       int sqlType = metadata.getColumnType(i + 1);
       int sqlPrecision = metadata.getPrecision(i + 1);
       int sqlScale = metadata.getScale(i + 1);
+      // handles backwards compatibility
+      boolean handleDecimalAsDouble = false;
+      if (sqlType == Types.DECIMAL && outputSchema != null) {
+        Schema outputFieldSchema = field.getSchema();
+        if (outputFieldSchema != null) {
+          outputFieldSchema = outputFieldSchema.isNullable() ? outputFieldSchema.getNonNullable() : outputFieldSchema;
+          handleDecimalAsDouble = outputFieldSchema.getType() == Schema.Type.DOUBLE;
```
Comment (CuriousVini) [inline database-plugins/src/main/java/io/cdap/plugin/DBRecord.java]: This does not cover all cases. The schema type changes depending on precision and scale values:
```
type = scale != 0 ? Schema.Type.DOUBLE :
          // with 10 digits we can represent 2^32 and LONG is required
          precision > 9 ? Schema.Type.LONG : Schema.Type.INT;

**Review by CuriousVini — CHANGES_REQUESTED** [review]

**Comment by CuriousVini** [comment]
Also please add tests to cover all the schema types derived for decimal type to cover backwards compatibility

File: database-plugins/src/main/java/io/cdap/plugin/DBRecord.java (hunk 5 above)
Comment (vjosadika) [reply database-plugins/src/main/java/io/cdap/plugin/DBRecord.java]: @CuriousVini fixed

File: database-plugins/src/test/java/io/cdap/plugin/db/batch/source/DBSourceTestRun.java [hunk 6]
```diff
@@ -351,8 +416,8 @@ public void testNonExistentDBTable() throws Exception {
       .addConnection(sourceBadName.getName(), sink.getName())
       .build();
     ApplicationId appId = NamespaceId.DEFAULT.app("dbSourceNonExistingTest");
-    assertDeploymentFailure(appId, etlConfig, "ETL Application with DB Source should have failed because of a " +
-      "non-existent source table.");
+    assertDeploymentFailure(appId, etlConfig, "ETL Application with DB Source should have failed because "
+      + "of a " + "non-existent source table.");
```
Comment (CuriousVini) [inline database-plugins/src/test/java/io/cdap/plugin/db/batch/source/DBSourceTestRun.java]: not needed

File: database-plugins/src/test/java/io/cdap/plugin/db/batch/source/DBSourceTestRun.java [hunk 7]
```diff
@@ -374,7 +439,7 @@ public void testNonExistentDBTable() throws Exception {
       .addStage(sink)
       .addConnection(sourceBadConn.getName(), sink.getName())
       .build();
-    assertDeploymentFailure(appId, etlConfig, "ETL Application with DB Source should have failed because of a " +
-      "non-existent source database.");
+    assertDeploymentFailure(appId, etlConfig, "ETL Application with DB Source should have failed because "
+      + "of a " + "non-existent source database.");
```
Comment (CuriousVini) [inline database-plugins/src/test/java/io/cdap/plugin/db/batch/source/DBSourceTestRun.java]: Is this unintended?

File: database-plugins/src/test/java/io/cdap/plugin/db/batch/source/DBSourceTestRun.java [hunk 8]
```diff
@@ -298,7 +362,8 @@ public void testUserNamePasswordCombinations() throws Exception {
     // as source
     Map<String, String> noUser = new HashMap<>(baseSourceProps);
     noUser.put(DBConfig.PASSWORD, "password");
-    database = new ETLStage("databaseSource", new ETLPlugin("Database", BatchSource.PLUGIN_TYPE, noUser, null));
+    database = new ETLStage("databaseSource", new ETLPlugin("Database", BatchSource.PLUGIN_TYPE, noUser,
```
Comment (CuriousVini) [inline database-plugins/src/test/java/io/cdap/plugin/db/batch/source/DBSourceTestRun.java]: same here. unintended?

File: database-plugins/src/main/java/io/cdap/plugin/DBUtils.java [hunk 9]
```diff
@@ -149,21 +149,21 @@ public static DriverCleanup ensureJDBCDriverIsAvailable(Class<? extends Driver>
...
-   * {@link DBUtils#getSchema(String, int, int, int, String)}
+   * {@link DBUtils#getSchema(String, int, int, int, String, boolean)}
    *
    * @param resultSet result set of executed query
    * @return list of schema fields
    * @throws SQLException
    */
-  public static List<Schema.Field> getOriginalSchema(ResultSet resultSet) throws SQLException {
-    return getSchemaFields(resultSet, null, null);
+  public static List<Schema.Field> getOriginalSchema(ResultSet resultSet, Schema outputSchema) throws SQLException {
```
Comment (CuriousVini) [inline database-plugins/src/main/java/io/cdap/plugin/DBUtils.java:158]: Mark `Schema outputSchema` as nullable

Comment (CuriousVini) [reply database-plugins/src/main/java/io/cdap/plugin/DBUtils.java:158]: Javadoc should be updated and should have same params as number of params in the method

File: database-plugins/src/main/java/io/cdap/plugin/DBUtils.java [hunk 10]
```diff
@@ -172,7 +172,8 @@ public static DriverCleanup ensureJDBCDriverIsAvailable(Class<? extends Driver>
    * @throws SQLException
    */
   public static List<Schema.Field> getSchemaFields(ResultSet resultSet, @Nullable String patternToReplace,
-                                                   @Nullable String replaceWith) throws SQLException {
+                                                   @Nullable String replaceWith, Schema outputSchema)
```
Comment (CuriousVini) [inline database-plugins/src/main/java/io/cdap/plugin/DBUtils.java]: Mark outpuschema as nullable

File: database-plugins/src/main/java/io/cdap/plugin/DBUtils.java [hunk 11]
```diff
@@ -232,13 +247,21 @@ private static Schema getSchema(String typeName, int sqlType, int precision, int
         break;
 
       case Types.NUMERIC:
```
Comment (CuriousVini) [inline database-plugins/src/main/java/io/cdap/plugin/DBUtils.java:249]: What is the difference between numeric and decimal type? Does this change need to be applied to numeric too?

File: database-plugins/src/main/java/io/cdap/plugin/DBUtils.java [hunk 12]
```diff
@@ -232,13 +247,21 @@ private static Schema getSchema(String typeName, int sqlType, int precision, int
...
           // with 10 digits we can represent 2^32 and LONG is required
           precision > 9 ? Schema.Type.LONG : Schema.Type.INT;
         break;
-
+      case Types.DECIMAL:
+        if (handleAsDecimal) {
+          return Schema.decimalOf(precision, scale);
+        } else {
+          // if there are no digits after the point, use integer types
+          type = scale != 0 ? Schema.Type.DOUBLE :
```
Comment (CuriousVini) [inline database-plugins/src/main/java/io/cdap/plugin/DBUtils.java:255]: should not have duplicate code. Please refactor it.

File: database-plugins/src/main/java/io/cdap/plugin/DBUtils.java [hunk 13]
```diff
@@ -293,6 +316,23 @@ public static Object transformValue(int sqlType, int precision, int scale,
           } else {
             return decimal.intValue();
           }
+        }
+        case Types.DECIMAL: {
+          if (handleAsDecimal) {
+            return original;
+          } else {
+            BigDecimal decimal = (BigDecimal) original;
+            if (scale != 0) {
```
Comment (CuriousVini) [inline database-plugins/src/main/java/io/cdap/plugin/DBUtils.java:309]: shouldnt duplicate the logic. Please refactor it.

File: database-plugins/src/main/java/io/cdap/plugin/DBRecord.java [hunk 14]
```diff
@@ -124,23 +137,37 @@ public void readFields(ResultSet resultSet) throws SQLException {
...
   }
 
   private void setField(ResultSet resultSet, StructuredRecord.Builder recordBuilder, Schema.Field field, int sqlType,
-                        int sqlPrecision, int sqlScale, String originalName) throws SQLException {
+                        int sqlPrecision, int sqlScale, String originalName, boolean handleAsDecimal)
+    throws SQLException {
     // original name has to be used to get result from result set
-    Object o = DBUtils.transformValue(sqlType, sqlPrecision, sqlScale, resultSet, originalName);
+    Object o = DBUtils.transformValue(sqlType, sqlPrecision, sqlScale, resultSet, originalName,
+                                      handleAsDecimal);
```
Comment (CuriousVini) [inline database-plugins/src/main/java/io/cdap/plugin/DBRecord.java]: There is no need for `handleAsDecimal ` variable in this method. We can just pass `Schema.Field field` to the method and get the type of field to derive if decimal type needs to be handled.

**Review by CuriousVini — CHANGES_REQUESTED** [review]
1. Has this been tested for database sink as well? 
2. What is the difference between numeric and decimal types in sql? 
3. Was backwards compatibility tested manually? If so, How was backwards compatibility tested?

**Comment by vjosadika** [comment]
> 1. Has this been tested for database sink as well?
> 2. What is the difference between numeric and decimal types in sql?
> 3. Was backwards compatibility tested manually? If so, How was backwards compatibility tested?

@CuriousVini 
1. Has this been tested for database sink as well?
Yes, this has been tested with Database Sink. The source was a Database that has a field of type decimal, linked with a Database Sink which gave a successful result. All unit tests were successful as well.
2. What is the difference between numeric and decimal type? Does this change need to be applied to numeric too?
Decimal and numeric are handled the same.
3. Was backwards compatibility tested manually? If so, How was backwards compatibility tested?
Backwards compatibility was tested manually. 
Test 1 - Database source with a schema that has a field of decimal type (5,2), linked with a record duplicator action plugin and a Database sink plugin. The result was successful.
Test 2- Exported pipeline from test 1. Imported it again and change the type of the field from decimal to double. The result was successful.
Test 3- Changed the field of decimal type(5,2) in mysql database to decimal (10,0). Exported pipeline from test 1. Imported it again and change the type of the field from decimal too long. The result was successful.
Test 4- Changed the field of decimal type(5,2) in mysql database to decimal (8,0). Exported pipeline from test 1. Imported it again and change the type of the field from decimal to int. The result was successful.

**Comment by CuriousVini** [comment]
> > 1. Has this been tested for database sink as well?
> > 2. What is the difference between numeric and decimal types in sql?
> > 3. Was backwards compatibility tested manually? If so, How was backwards compatibility tested?
> 
> @CuriousVini
> 
> 1. Has this been tested for database sink as well?
>    Yes, this has been tested with Database Sink. The source was a Database that has a field of type decimal, linked with a Database Sink which gave a successful result. All unit tests were successful as well.
> 2. What is the difference between numeric and decimal type? Does this change need to be applied to numeric too?
>    Decimal and numeric are handled the same.
> 3. Was backwards compatibility tested manually? If so, How was backwards compatibility tested?
>    Backwards compatibility was tested manually. 
>    Test 1 - Database source with a schema that has a field of decimal type (5,2), linked with a record duplicator action plugin and a Database sink plugin. The result was successful.
>    Test 2- Exported pipeline from test 1. Imported it again and change the type of the field from decimal to double. The result was successful.
>    Test 3- Changed the field of decimal type(5,2) in mysql database to decimal (10,0). Exported pipeline from test 1. Imported it again and change the type of the field from decimal too long. The result was successful.
>    Test 4- Changed the field of decimal type(5,2) in mysql database to decimal (8,0). Exported pipeline from test 1. Imported it again and change the type of the field from decimal to int. The result was successful.

For 
1. The output in database was written as decimal? 
2. I see
3. sounds good 👍

**Review by CuriousVini — APPROVED** [review]
lgtm 👍 
Please squash the commits

**Review by CuriousVini — COMMENTED** [review]
The output in database was written as decimal?

**Comment by vjosadika** [comment]
> The output in database was written as decimal?

Yes

**Pull request merged**